crates/holochain_integrity_types/src/countersigning.rs

```diff
@@ -184,28 +198,52 @@ impl PreflightRequest {
         }
     }
 
+    /// Verify the optional signing agents.
+    pub fn check_agents_optional(&self) -> Result<(), CounterSigningError> {
+        if self.minimum_optional_signing_agents as usize > self.optional_signing_agents.len() {
+            return Err(CounterSigningError::OptionalAgentsLength(
+                self.minimum_optional_signing_agents,
+                self.optional_signing_agents.len(),
+            ));
+        }
+        // Minimum optional signers must be at least half the total signers.
+        if ((self.minimum_optional_signing_agents * 2) as usize)
+            < self.optional_signing_agents.len()
+            && !self.optional_signing_agents.is_empty()
+        {
+            return Err(CounterSigningError::MinOptionalAgents(
+                self.minimum_optional_signing_agents,
+                self.optional_signing_agents.len(),
+            ));
+        }
+        Ok(())
+    }
+
     /// Verify the preflight request agents.
     pub fn check_agents(&self) -> Result<(), CounterSigningError> {
         self.check_agents_dupes()?;
         self.check_agents_len()?;
+        self.check_agents_optional()?;
         Ok(())
     }
 
-    /// Verify the enzyme index is in bounds of the signing agent if set.
-    pub fn check_enzyme_index(&self) -> Result<(), CounterSigningError> {
-        match self.enzyme_index() {
-            Some(index) => {
-                if (*index as usize) < self.signing_agents().len() {
-                    Ok(())
-                } else {
-                    Err(CounterSigningError::EnzymeIndex(
-                        self.signing_agents().len(),
-                        *index as usize,
-                    ))
-                }
-            }
-            None => Ok(()),
+    /// Verify everything about the enzyme.
+    pub fn check_enzyme(&self) -> Result<(), CounterSigningError> {
+        // Enzymatic optional signing agents MUST match the first signer in
+        // both the signing agents and optional signing agents.
+        if self.enzymatic
+            && !self.optional_signing_agents.is_empty()
+            && self.signing_agents[0] != self.optional_signing_agents[0]
```

Comment:
Could this 0 ever be out of bounds. Maybe it's better to use `.get(0)` instead of array indexing for all these checks? Otherwise you get a panic